### PR TITLE
feat: resolve data overlay files on the take (and scan) read path

### PR DESCRIPTION
**Pull request by wjones127**
> **Supersedes #7409.** Relocated into `lance-format/lance` and **properly stacked on the OSS-1322 branch** so the diff is now exactly this PR's change (no spec/1322/1323 commits to scroll past).

Resolves the `take` random-access path against data overlay files, replacing the temporary "overlays not supported" error from OSS-1322. Implements OSS-1324.

Because `take` and scan share `FragmentReader`'s read path, the merge is wired there once: each row is addressed by its physical offset (from `ReadBatchParams::to_offsets_total`) and resolved against the overlays that cover its field. This also enables the scan-path merge that the OSS-1322 PR stubbed out.

### How it works
- **Lazy, rank-pushed reads.** Each contributing overlay file is opened once (projected to the covered ∩ requested fields) and **no value bytes are read up front**. Requested offsets are routed to the newest covering overlay at their coverage rank, and only the touched ranks are fetched — via the existing `take` primitive — so overlay reads are O(requested rows), not O(coverage).
- **Concurrent IO.** Base and overlay reads are issued together, then the column is assembled with `interleave` (rank → fetch-position remap).
- The merge runs on physical rows in read order, **before** deletion filtering, so:
  - deletions take precedence (an overlay value computed for a deleted row is dropped with the row),
  - NULL overrides apply (a covered offset with a NULL value resolves to NULL, distinct from fall-through),
  - fields resolve independently.
- Sparse per-field overlays read each field's value column independently, so unequal-length value columns (OSS-1323) need no rectangular batch. Rank-based addressing only (rank on the coverage bitmap + a value fetch; no offset key column, no binary search).

Overlays on nested (non-top-level) fields are not yet matched and are left for follow-up. There's a `TODO(overlay perf)` on reader priority to settle with a benchmark.

### Tests
take covered/uncovered offsets; multiple overlays (newest wins); per-field coverage with unequal-length columns; NULL override; overlay on a deleted row (inert); multi-fragment scan — each over v2.0 and v2.1. Plus unit tests for the routing/assembly core, and an IO guard (`test_take_reads_only_needed_overlay_ranks`) asserting a 2-row take over a fully-covering 100k-row overlay reads ~one miniblock, not the whole value column.

### Stacking
Stacked on the **OSS-1322** branch (`will/oss-1322-write-data-overlay-and-scan-data`). Merge that first; this PR's base retargets to `main` automatically when it lands.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for reading dataset fragments that include overlay files.
  * Overlay values are resolved and merged into results during both scans and targeted row reads, honoring newest-wins semantics across nested fields.
  * Optimizes reads by fetching only overlay data needed for the requested rows and projections.

* **Bug Fixes**
  * Fragments with overlays are no longer rejected, and overlay resolution remains correct when overlays overlap deleted rows.

* **Tests**
  * Added end-to-end overlay read coverage for scan/take, including multi-batch slicing and edge cases.

### CODE REVIEW

**Comment by wjones127** [comment]
## Read impact of overlay files

What does resolving overlays on read cost, vs. a dataset with none? Measured on a
1M-row `int32` column, 1% of cells overlaid, local SSD, warm cache. Each layer is
a separate overlay file all covering the same cells (the worst case — every layer
must be consulted). `take` is 100 random rows; `scan` is a full scan of the
column.

| coverage | overlays | take (ms) | scan (ms) |
|---|---:|---:|---:|
| — | **0 (none)** | 0.5 | 11 |
| contiguous | 4  | 0.4 | 11 |
| contiguous | 16 | 0.5 | 21 |
| stride     | 4  | 0.5 | 19 |
| stride     | 16 | 0.5 | 35 |

**`take` is unaffected by overlays.** Rank pushdown reads only the overlaid cells
a `take` actually touches, so a 100-row `take` costs the same (~0.5 ms) with 0 or
16 layers.

**`scan` pays a per-layer merge cost.** A full scan grows roughly linearly with
the number of layers, since each layer adds one value-column read and one
`interleave` pass over the output. Scattered (stride) coverage costs more than a
single contiguous run (more pages touched, larger/denser coverage bitmaps): ~1.5
ms/layer strided vs. ~0.6 ms/layer contiguous here. A small number of layers over
contiguous regions is nearly free; many layers — especially scattered — is where
cost accrues, and is exactly what overlay compaction folds back down.

The routing itself (deciding base-vs-overlay per cell) is negligible — under 2 ms
even at 16 scattered layers — so scan cost is dominated by the actual value reads
and merges, not bookkeeping.

**Comment by wjones127** [comment]
## Read impact of overlay files on a **wide** value column

Follow-up to the `int32` numbers [above](https://github.com/lance-format/lance/pull/7536#issuecomment-4848828467). That table used a 4-byte `int32` column, where the base read is nearly free — but ML overlays usually target wide vectors, where the base read is the whole story. So here is the same take/scan sweep on a **3072-d `float32` embedding** (12 KiB/row, ~750× an `int32` cell). 100K rows (base file ~1.2 GiB), 1% of cells overlaid, single fragment, local SSD, warm cache; each layer covers the same cells (worst case — every layer must be consulted). Both dtypes re-measured on the same machine so they're comparable. Median wall time:

### Wide column — 3072-d embedding, 100K rows (~1.2 GiB base)

| coverage | overlays | take (ms) | scan (ms) |
|---|---:|---:|---:|
| — | **0 (none)** | 0.4 | ~600 |
| contiguous | 4  | 0.5 | 595 |
| contiguous | 16 | 0.6 | 592 |
| stride     | 4  | 0.5 | 856 |
| stride     | 16 | 0.6 | 858 |

For reference, the narrow `int32` column (1M rows) re-run on the same machine: scan **12 / 14 / 27 ms** contiguous at 0 / 4 / 16 layers, **13 / 35 / 41 ms** stride; take **0.3 – 0.5 ms** throughout — same shape as the original table.

**Takeaways for a wide column:**

**`take` stays ~free regardless of width.** A 100-row `take` is sub-millisecond even at 3072-d, because rank pushdown reads only the ~1.2 MB of cells the take actually touches. Column width barely moves it, and the per-layer cost is the same small additive bump as `int32` (0.4 → 0.6 ms from 0 → 16 layers).

**`scan` is dominated by the base column read, not overlay bookkeeping.** Reading and decoding the 1.2 GiB base embedding is ~600 ms — that's the floor every scan pays before a single overlay is considered.

**Contiguous overlays are effectively free on a wide scan.** 0 / 4 / 16 contiguous layers all land at ~590–620 ms, no measurable per-layer cost — the bitmap-major contiguous path from this PR doing exactly what it's for: against a base read that large, the contiguous merge is lost in the noise. (Contrast `int32`, where the base read is trivial, so each contiguous layer is plainly visible: 12 → 14 → 27 ms.)

**Scattered (stride) coverage adds a bounded ~45% (~270 ms) that — unlike `int32` — does not grow with layer count.** 4 and 16 stride layers are indistinguishable (both ~856 ms). Once a scan is on the scattered-merge path, the extra cost is a one-time scattered gather over wide rows; stacking more layers over the same cells only unions cheap coverage bitmaps while the base read still dominates. On `int32` the per-layer merge *is* the whole cost, so it keeps climbing (35 → 41 ms).

Cold-cache, every wide scan reads the full ~1.2 GiB column regardless of layer count (16 layers at 1% coverage add <5% IO); a wide `take` reads ~1.2 MB.

**Net:** for wide vector columns the overlay read tax is smaller in *relative* terms than the `int32` microbenchmark suggests — contiguous layers are essentially free, and even scattered layers cost a bounded ~45% that doesn't compound with depth. The `int32` sweep is close to a worst case for relative overhead precisely because its base read is so cheap there's nothing to amortize against.

<sub>Benchmark added in #7544 (`test_overlay_read_wide`).</sub>

File: rust/lance/src/dataset/fragment.rs [hunk 1]
```diff
@@ -2604,6 +2619,50 @@ impl FragmentReader {
...
+        &self,
+        merged: ReadBatchTaskStream,
+        params: &ReadBatchParams,
+        total_num_rows: u32,
+    ) -> ReadBatchTaskStream {
+        if self.overlay_plans.is_empty() {
+            return merged;
+        }
+        let offsets: Arc<Vec<u32>> =
+            Arc::new(params.to_offsets_total(total_num_rows).values().to_vec());
```
Comment (wjones127) [inline rust/lance/src/dataset/fragment.rs]: question: how expensive is this?

File: rust/lance/src/dataset/overlay.rs [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+/// individual cells. The in-range bits then carry **consecutive** coverage ranks
+/// starting at the count of bits below `base` (one `rank` lookup) — no bits lie
+/// between them by construction — so ranks need no running count. Coverages are
+/// processed newest-first with a "first claim wins" guard for precedence.
+fn route_contiguous(
+    base: u32,
+    len: usize,
+    coverages_newest_first: &[&RoaringBitmap],
+) -> OverlayRouting {
+    let mut needed_ranks: Vec<Vec<u32>> = vec![Vec::new(); coverages_newest_first.len()];
```
Comment (wjones127) [inline rust/lance/src/dataset/overlay.rs]: suggestion: I honestly find the language of "rank" confusing. I think what I really want is to label "offset_in_overlay", "offset_in_batch", "offset_in_frag". IMO that's easier to reason about.

File: rust/lance/src/dataset/overlay.rs [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+
+    // Open each contributing overlay file once, concurrently. The reader is
+    // shared (via `Arc`) by every field that file covers.
+    //
+    // TODO(overlay perf): these reads use the default reader priority. Once
+    // we benchmark take/scan over overlays, decide whether overlay value
+    // reads should inherit `read_config.reader_priority` (or get a dedicated
+    // priority) so they schedule alongside the base reads.
+    let opened: Vec<Option<Arc<dyn GenericFileReader>>> =
+        futures::future::try_join_all(order.iter().map(|&overlay_idx| async move {
```
Comment (wjones127) [inline rust/lance/src/dataset/overlay.rs]: question: at this point have we pruned irrelevant overlay readers? Ones that don't overlap our projection or our row selection?

File: rust/lance/src/dataset/overlay.rs [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+pub fn overlay_indices_newest_first(overlays: &[DataOverlayFile]) -> Vec<usize> {
+    let mut indices: Vec<usize> = (0..overlays.len()).collect();
+    indices.sort_by(|&a, &b| {
+        overlays[b]
+            .committed_version
+            .cmp(&overlays[a].committed_version)
+            .then(b.cmp(&a))
+    });
+    indices
+}
```
Comment (wjones127) [inline rust/lance/src/dataset/overlay.rs]: suggestion: slices have a double-ended iterator, so why not just read the slices as-is in reverse order?

File: rust/lance/src/dataset/overlay.rs [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+                .open_reader(&overlay.data_file, Some(&schema), read_config)
+                .await?
+                .map(Arc::from))
+        }))
+        .await?;
+
+    let mut plans = Vec::new();
+    for field in &projection.fields {
+        let mut overlays_newest_first = Vec::new();
+        for (slot, &overlay_idx) in order.iter().enumerate() {
```
Comment (wjones127) [inline rust/lance/src/dataset/overlay.rs]: issue(blocking): this is `O(num_fields x num_overlays)`, where `num_fields` could be thousands of fields. Can we instead create an index based on the overlays in `order`, and use that to quickly find the relevant fields?

File: rust/lance/src/dataset/overlay.rs [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+/// lists, so planning is `O(total overlay fields + projected fields)` rather than
+/// `O(projected fields × overlays)` (a fragment can have thousands of fields).
+///
+/// An overlay contributes to a projected field when any id in its `data_file.fields`
+/// belongs to that top-level field's subtree. Overlays are written against the field
+/// ids the file actually stores, which for a struct or list column is its *leaf*
+/// child ids (the V2_1 structural encoding does not record the parent id), so a leaf
+/// is mapped back to its top-level projected ancestor — the whole column is then
+/// fetched and replaced as a unit. Each contributing overlay *file* appears once in
+/// `files`, shared by every top-level field it covers.
```
Comment (westonpace) [inline rust/lance/src/dataset/overlay.rs]: The wording here is confusing.

> An overlay contributes to a projected field when any id in its `data_file.fields` belongs to **that** top-level field's subtree

What "top-level field" is the emphasized 'that' referring to?  I think it is referring to the projected field but projected fields are not necessarily top-level fields.

File: rust/lance/src/dataset/overlay.rs [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+        // as whole-value replacement shares one coverage across the leaves).
+        let mut seen = HashSet::new();
+        let mut contributions: Vec<(&Field, usize)> = Vec::new();
+        for (field_pos, &field_id) in overlay.data_file.fields.iter().enumerate() {
+            if let Some(&top) = top_level_of.get(&field_id)
+                && seen.insert(top.id)
+            {
+                contributions.push((top, field_pos));
+            }
+        }
```
Comment (westonpace) [inline rust/lance/src/dataset/overlay.rs:421]: Let say I have a struct as so:

```
my_struct(field_id=7)
my_struct.foo(field_id=8)
my_struct.bar(field_id=9)
```

If my projection targets field 8 and the overlay targets field 9 then I will not find it here and the overlay will not be applied.  I think this is correct but all this talk of top-level fields made it seem like an overlay should be applied whenever there is a common top-level field (which there is in this case).

Comment (westonpace) [reply rust/lance/src/dataset/overlay.rs:421]: Actually, here's a more concerning case.  Let's say I have:

```
outer_struct(field_id=11)
outer_struct.middle_struct(field_id=12)
outer_struct.middle_struct.inner_struct(field_id=13)
```

Let's say my projection asks for `field_id=12` (middle_struct).  Then `top_level_of` will be `12->11`.  Then my overlay will have `field_id=13` which is not present in `top_level_of` and so I will not apply that overlay _but I should_!

File: rust/lance/src/dataset/fragment.rs [hunk 8]
```diff
@@ -2604,6 +2629,72 @@ impl FragmentReader {
         Ok(result.project_by_schema(&output_schema)?)
     }
 
+    /// Merge data overlay values into a stream of base batches.
```
Comment (westonpace) [inline rust/lance/src/dataset/fragment.rs]:
```suggestion
    /// Merge data overlay values onto a stream of base batches.
```
Super trivial and possibly incorrect nit.

File: rust/lance/src/dataset/fragment.rs [hunk 9]
```diff
@@ -2604,6 +2629,72 @@ impl FragmentReader {
...
+    ) -> Result<ReadBatchTaskStream> {
+        let Some(overlay) = &self.overlay else {
+            return Ok(merged);
+        };
+        // The offset_in_frag of every row this read will return, materialized once.
+        // Cost is one u32 per output row (a whole-fragment scan is 4 bytes/row), and
+        // it lets us both prune overlays to the read and slice each batch's offsets
+        // below without reading any data. Only paid when the fragment has overlays.
+        let offsets_in_frag: Arc<Vec<u32>> =
+            Arc::new(params.to_offsets_total(total_num_rows).values().to_vec());
```
Comment (westonpace) [inline rust/lance/src/dataset/fragment.rs:2676]: I wonder if we can avoid this materialization?  In `resolve_overlays` it is immediately converted into a bitmap.  In the map function it is sliced.  These seem like things we could do without materializing.  I think slicing support already exists for `ReadBatchParams` and we could add conversion-to-bitmap support (or just convert to bitmap immediately and then slice the bitmap)

Alternatively, maybe `ReadBatchParams` should change to a roaring bitmap/treemap in the first place 🤔 

But I guess that would be a larger refactor 😛

File: rust/lance/src/dataset/overlay.rs [hunk 10]
```diff
@@ -0,0 +1,848 @@
...
+    }
+
+    // Open each needed file once, concurrently. The reader is shared (via `Arc`) by
+    // every field that file covers.
+    //
+    // TODO(overlay perf): these reads use the default reader priority. Once we
+    // benchmark take/scan over overlays, decide whether overlay value reads should
+    // inherit `read_config.reader_priority` (or get a dedicated priority) so they
+    // schedule alongside the base reads.
```
Comment (westonpace) [inline rust/lance/src/dataset/overlay.rs]: It's more than just a performance concern.  An invalid priority can lead to deadlock.  For example, the default prioirty is 0 (highest priority), if we issue a bunch of overlay reads for a batch that we're not ready to read at all then they can clog up the backpressure queue.

However, I _think_ we are ok.  I don't think we initiate the fetch until the task is polled for the first time.  That won't happen unless we have a dedicated consumer for the data.  A priority of 0 is probably the correct answer in this case (we are initiating the reads quite late in the process)

The potential performance optimization would be to start fetching the overlay data earlier.  For example, imagine we are in a compute constrained workload.  The current approach probably gives us bubbles:

Image

So, in the end, I guess it is a perf question.  Still, let's update the comment to say we are using priority 0 because we've already indicated we are ready to consume this batch and so it is safe to consider it highest priority.

File: rust/lance/src/dataset/overlay.rs [hunk 11]
```diff
@@ -0,0 +1,848 @@
...
+/// this a block at a time, so a coverage that does not overlap the batch (e.g. a
+/// scan batch past the last cell this overlay touches) is skipped cheaply without
+/// inspecting individual bits.
+///
+/// Within the batch a coverage's cells appear in ascending order, so their
+/// `offset_in_overlay` values are consecutive: the first in-batch cell sits at
+/// `offset_in_overlay = <cells this coverage has before the batch>` (a single
+/// `rank` lookup), and each following cell is one more. Coverages are applied
+/// newest-first, and the first overlay to claim a row wins.
+fn route_contiguous(
```
Comment (westonpace) [inline rust/lance/src/dataset/overlay.rs:124]: Nice fast path

**Review by westonpace — APPROVED** [review]
Clean design.  I think the I/O could probably be optimized better in the future (I dropped a comment about the holes) but we can cross that bridge when we come to it.

**Review by westonpace — CHANGES_REQUESTED** [review]
Actually, sorry, switching back to request changes until the projection mapping comment (https://github.com/lance-format/lance/pull/7536#discussion_r3553495147) is resolved.

File: rust/lance/src/dataset/fragment.rs [hunk 12]
```diff
@@ -3114,6 +3217,1519 @@ mod tests {
...
+                .downcast_ref::<StructArray>()
+                .unwrap();
+            // a: offset 1 base (1), offset 2 overlaid (777); b untouched.
+            assert_eq!(i32_child(middle, 0).values(), &[1, 777]);
+            assert_eq!(i32_child(middle, 1).values(), &[100, 200]);
+
+            // Projecting the *intermediate* struct `outer.middle` (field id 2) while
+            // the overlay targets a deeper field (id 3) must still apply: the
+            // overlay's leaf id falls inside the projected subtree, so it maps to a
+            // projected atomic field. (This is the case wjones127/westonpace flagged where a
```
Comment (westonpace) [inline rust/lance/src/dataset/fragment.rs:4720]: Some of these test comments are probably a little superfluous but I know how hard it is to avoid generating them 😆

**Review by westonpace — APPROVED** [review]
Thanks for testing the extra cases.  This covers the possibilities I can think of.  Any others we can address as we find them.
